rfcs/v3/authorization-rules.md

```diff
@@ -0,0 +1,568 @@
+  commandName: DeleteReviews
+  authorizationRules:
+    # Rule 1: Allow users to only delete their own reviews
+    - presetArguments:
+      - argumentName: restriction
+        value:
+          includeBooleanExpression:
+            fieldComparison:
+              fieldName: user_id
+              operator: _eq
+              value:
+                sessionVariable: x-hasura-user-id
+      condition:
+          equals:
+            left: sessionVariable: x-hasura-role
+            right: literal: user
+    # Rule 2: Do not allow anyone except admins to delete flagged reviews
+    - presetArguments:
+      - argumentName: restriction
+        value:
+          excludeBooleanExpression:
+            fieldComparison:
+              fieldName: is_flagged
+              operator: _eq
+              value:
+                literal: true
+      condition:
+          not:
+            equals:
+              left: sessionVariable: x-hasura-role
+              right: literal: admin
+```
+
+### Reusing Relationship Permissions
+
+In the existing authorization system, the entire permissions predicate needs to be specified for a model's permission. For related models, however, this can result in a duplicate and very nested permission predicates.
+
+For example, if I have models for `Channels`, `Threads`, `Replies`. Then my predicates would be as follows:
+- `Channels`: `members._contains(x-hasura-user-id)`
+- `Threads`: `channel.members._contains(x-hasura-user-id`
+- `Replies`: `thread.channel.members._contains(x-hasura-user-id)`.
+
+Ideally, you would be able to express the fact that an object should be selectable if the related object is also selectable.
+
+Hence, we propose the following addition to the permissions predicate:
+```yaml
+allowObjects:
+  relationship:
+    name: 'my_relationship'
+    relatedObjectAllowed: true
+```
+
+### AuthConfig
+
+The notion of roles will be removed from AuthConfig. In particular, the `allow_role_emulation_by` key will be replaced by a `allow_session_emulation_if` key which specifies a condition which needs to evaluate to true for the request to allow emulating session variables.
+
+No `x-hasura-` session variables will now be required in the webhook response or in JWT claims. Users can choose arbitrary names for session variables.
+
+### Implementation
+
+The above proposal has major implications for engine, the `lang-graphql` crate in particular. Currently, `lang-graphql` allows constructing a single graphql schema with information about all roles embedded into the schema and allows query validation / IR generation to happen in the context of a particular role only. However, with the above proposal, there is no notion of roles anymore. So, `lang-graphql` will need to move to an architecture where when trying to access a graphql field, instead of looking up an allowed set of roles, it can evaluate an arbitrary expression that evaluates to a boolean. The details of this architecture are out of scope for this RFC.
```

Comment:
This presumably means losing the ability to generate a GraphQL schema (and thus generate types etc) for only the queries / types that a given role is allowed to access. This feels like quite a price to pay, are we satisfied that users don't need this?

Reply:
No, we would still generate the GraphQL schema only according to what's accessible to a given set of session variables.

When an introspection query comes in, we would evaluate which fields/types are accessible based on the session variables associated with the request and generate a schema containing only those types/fields.